src/Validation/Combinators.hs

```diff
@@ -0,0 +1,211 @@
+{- |
+Copyright:  (c) 2020 Kowainik
+SPDX-License-Identifier: MPL-2.0
+Maintainer: Kowainik <xrom.xkov@gmail.com>
+
+Helpful combinators to work with 'Validation' data type.
+-}
+
+module Validation.Combinators
+    ( validateAll
+
+      -- * When* functions
+    , whenSuccess
+    , whenFailure
+    , whenSuccess_
+    , whenFailure_
+    , whenSuccessM
+    , whenFailureM
+    , whenSuccessM_
+    , whenFailureM_
+
+      -- * 'Maybe' conversion
+    , failureToMaybe
+    , successToMaybe
+    , maybeToFailure
+    , maybeToSuccess
+
+    ) where
+
+import Data.Foldable (foldl')
+
+import Validation (Validation (..), validation)
+
+
+{- | Validate all given checks in a 'Foldable'. Returns the 'Success' of the
+start element when all checks are successful.
+
+
+A basic example of usage could look like this:
+
+@
+> let validatePassword = validateAll
+        [ validateEmptyPassword
+        , validateShortPassword
+        ]
+
+> validateAll "VeryStrongPassword"
+Success "VeryStrongPassword"
+
+> validateAll ""
+Failure (EmptyPassword :| [ShortPassword])
+@
+-}
+validateAll :: (Foldable f, Semigroup e) => f (a -> Validation e b) -> a -> Validation e a
```

Comment:
I think we can use explicit `forall` here to put `e` on the first place. I imagine people can use polymorphic functions over `Semigroup`, like `one` from `relude` to create `NonEmpty` lists, so it could improve API to be able to specify `Semigroup` easier.